compiler/rustc_middle/src/ty/mod.rs

```diff
@@ -463,16 +463,31 @@ static_assert_size!(WithStableHash<TyS<'_>>, 56);
 #[rustc_pass_by_value]
 pub struct Ty<'tcx>(Interned<'tcx, WithStableHash<TyS<'tcx>>>);
 
-// Statics only used for internal testing.
-pub static BOOL_TY: Ty<'static> = Ty(Interned::new_unchecked(&WithStableHash {
-    internee: BOOL_TYS,
-    stable_hash: Fingerprint::ZERO,
-}));
-const BOOL_TYS: TyS<'static> = TyS {
-    kind: ty::Bool,
-    flags: TypeFlags::empty(),
-    outer_exclusive_binder: DebruijnIndex::from_usize(0),
-};
+static LEAKED_BOOL_TY_ALREADY: std::sync::atomic::AtomicBool =
+    std::sync::atomic::AtomicBool::new(false);
+
+/// "Static" bool only used for internal testing.
+///
+/// FIXME(rustc_type_ir): This really should be replaced with something that doesn't leak.
+/// however, since it's used for testing, it's not _that_ bad.
+pub fn leak_bool_ty_for_unit_testing<'tcx>() -> Ty<'tcx> {
+    use std::sync::atomic::*;
+
+    if LEAKED_BOOL_TY_ALREADY.load(Ordering::Acquire) {
+        panic!("Can only leak one bool type, since its equality depends on its address");
+    } else {
+        LEAKED_BOOL_TY_ALREADY.store(true, Ordering::Release);
+    }
```

Comment:
Couldn't two threads both load `false` and then store `true`, thus creating two different values?

```suggestion
    if LEAKED_BOOL_TY_ALREADY.swap(true, Ordering::AcqRel) {
        panic!("Can only leak one bool type, since its equality depends on its address");
    }
```

Reply:
I'm probably going to get rid of this in favor of another solution, actually.